kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/storage/cloud/KieServerStateCloudRepository.java

```diff
@@ -0,0 +1,150 @@
+    private static final Logger logger = LoggerFactory.getLogger(KieServerStateCloudRepository.class);
+    private final XStream xs;
+
+    public static XStream initializeXStream() {
+        XStream xs = XStreamUtils.createTrustingXStream(new PureJavaReflectionProvider());
+        String[] voidDeny = {"void.class", "Void.class"};
+        xs.denyTypes(voidDeny);
+        xs.alias(CFG_MAP_DATA_KEY, KieServerState.class);
+        xs.alias("container", KieContainerResource.class);
+        xs.alias("config-item", KieServerConfigItem.class);
+
+        return xs;
+    }
+
+    public KieServerStateCloudRepository() {
+        xs = initializeXStream();
+    }
+
+    @Override
+    public void store(@NotNull String serverId, @NotNull KieServerState kieServerState) {
+
+        String kssServerId = null;
+        boolean stateUpdated = false;
+
+        try {
+            kssServerId = kieServerState.getConfiguration()
+                                        .getConfigItem(KieServerConstants.KIE_SERVER_ID).getValue();
+            KieServerConfigItem ksci = kieServerState.getConfiguration()
+                    .getConfigItem(KIE_SERVER_STATE_UPDATED);
+            
+            if (ksci != null) {
+                stateUpdated = true;
+                kieServerState.getConfiguration().removeConfigItem(ksci);
+            }
+        } catch (Exception e) {
+            logger.error("Failed to retrieve server id from  KieServerState", e);
+        }
+
+        if (kssServerId == null || serverId.length() == 0) {
+            throw new IllegalArgumentException("Invalid KieServerId: Can not be null or empty.");
+        }
+
+        if (!kssServerId.equals(serverId)) {
+            throw new IllegalArgumentException("Invalid KieServerId: Id does not match with KieServerState.");
+        }
+
+        try (KubernetesClient client = createAdaptiveClient()) {
+
+            ConfigMap cm = client.configMaps().withName(serverId).get();
+
+            if (cm == null) {
+                throw new IllegalStateException("KieServerState ConfigMap must exist before save/update.");
+            }
+
+            String stateXML = xs.toXML(kieServerState);
+            ObjectMeta md = cm.getMetadata();
+
+            Map<String, String> ann = md.getAnnotations() == null ? new ConcurrentHashMap<>() : md.getAnnotations();
+            md.setAnnotations(ann);
+            ann.put(STATE_CHANGE_TIMESTAMP,
+                    ZonedDateTime.now().format(DateTimeFormatter.ISO_INSTANT));
+            
+            if (stateUpdated) {
+                ann.put(ROLLOUT_REQUIRED, "true");
+            }
+
+            Map<String, String> cmData = new ConcurrentHashMap<>();
+            cmData.put(CFG_MAP_DATA_KEY, stateXML);
+            cm.setData(cmData);
+
+            client.configMaps().createOrReplace(cm);
+
+        } catch (IllegalStateException ise) {
```

Comment:
What is the reason for throwing IllegalStateException? I would expect the same behaviour as for other exceptions.

Reply:
The 'ise' is singling an un-recoverable exception for not being able to load the seeded ConfigMap for KieServer. (See below)

https://github.com/rhtevan/droolsjbpm-integration/blob/abf0dcad8de3b03909649fd1f67af4e0b4e1c0e5/kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/storage/cloud/KieServerStateCloudRepository.java#L95-L97

Reply:
https://github.com/kiegroup/droolsjbpm-integration/blob/abf0dcad8de3b03909649fd1f67af4e0b4e1c0e5/kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/storage/cloud/KieServerStateCloudRepository.java#L95-L97

Reply:
Ok, thanks.